python/dolfin/fem/form.py

```diff
@@ -1,85 +1,86 @@
+            self.form_compiler_parameters["external_include_dirs"] = dolfin_pc["include_dirs"]
 
-        ufc_form = ffc_jit(form, form_compiler_parameters=form_compiler_parameters,
+        # Extract subdomain data from UFL form
+        sd = form.subdomain_data()
+        self._subdomains, = list(sd.values())  # Assuming single domain
+        domain, = list(sd.keys())  # Assuming single domain
+        mesh = domain.ufl_cargo()
+
+        # Compile UFL form with JIT
+        ufc_form = ffc_jit(form, form_compiler_parameters=self.form_compiler_parameters,
                            mpi_comm=mesh.mpi_comm())
+        # Cast compiled library to pointer to ufc_form
         ufc_form = cpp.fem.make_ufc_form(ufc_form[0])
 
+        # For every argument in form extract its function space
         function_spaces = [func.function_space()._cpp_object for func in form.arguments()]
 
-        cpp.fem.Form.__init__(self, ufc_form, function_spaces)
+        # Prepare dolfin.Form and hold it as a member
+        self._cpp_object = cpp.fem.Form(ufc_form, function_spaces)
 
+        # Need to fill the form with coefficients data
+        # For every coefficient in form take its CPP object
         original_coefficients = form.coefficients()
-        self.coefficients = []
-        for i in range(self.num_coefficients()):
-            j = self.original_coefficient_position(i)
-            self.coefficients.append(original_coefficients[j].cpp_object())
+        for i in range(self._cpp_object.num_coefficients()):
+            j = self._cpp_object.original_coefficient_position(i)
+            self._cpp_object.set_coefficient(j, original_coefficients[i].cpp_object())
 
-        # Type checking coefficients
-        if not all(isinstance(c, (cpp.function.GenericFunction))
-                   for c in self.coefficients):
-            coefficient_error = "Error while extracting coefficients. "
-            raise TypeError(coefficient_error +
-                            "Either provide a dict of cpp.function.GenericFunctions, " +
-                            "or use Function to define your form.")
-
-        for i in range(self.num_coefficients()):
-            if isinstance(self.coefficients[i], cpp.function.GenericFunction):
-                self.set_coefficient(i, self.coefficients[i])
+        if mesh is None:
+            raise RuntimeError("Expecting to find a Mesh in the form.")
 
         # Attach mesh (because function spaces and coefficients may be
         # empty lists)
         if not function_spaces:
-            self.set_mesh(mesh)
+            self._cpp_object.set_mesh(mesh)
 
         # Attach subdomains to C++ Form if we have them
-        subdomains = self.subdomains.get("cell")
+        subdomains = self._subdomains.get("cell")
         if subdomains is not None:
-            self.set_cell_domains(subdomains)
-        subdomains = self.subdomains.get("exterior_facet")
+            self._cpp_object.set_cell_domains(subdomains)
```

Comment:
Should be no need to do `if subdomains is not None:`, etc. Just doing
```
self._cpp_object.set_cell_domains(subdomains)
``` 
should work fine in all cases. When `subdomains` is `None`, it will turn into `null` going into the C++ layer.